docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+            var additionalTexts = context.Sources.AdditionalTexts;
+
+            // apply a 1-to-1 transform on each text, extracting the path
+            var transformed = additionalTexts.Transform((text) => text.Path);
+
+            // batch the collected file paths into a single collection
+            var batched = transformed.BatchTransform(texts => texts);
+
+            // create an output that takes the file paths from the above batch and makes some user visible syntax
+            var output = batched.GenerateSource((sourceContext, filePaths) =>
+            {
+                sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}
+");
+            });
+
+            // register the output context.RegisterOutput(output); }); } }
+```
+
+## Advanced Implementation
+
+### Combining and filtering
+
+While the transformation steps outlined above allow a user to create simple
+generators from a single source of data, in reality it is expected that an
+author will need a way to take multiple sources of data and combine them.
+
+There exists an extension method `Combine` that allows the author to take one
+data source and merge it with another, for example, extracting a set of types
+from the compilation and using them to generate something on a per additional
+file basis.
+
+```csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // join 1 => many ((source1[0], source2), (source1[0], source2), ...)
+    public static IncrementalValueSource<(T, IEnumerable<U>)> Combine<T, U>(this IncrementalValueSource<T> source1, IncrementalValueSource<U> source2) => ...
+}
+```
+
+The second data source is batched, and the resulting step has a value type of
+`(T, IEnumerable<U>)`. That is, a tuple where each element consists of a single
+item of data from `source1` combined with the entire batch of data from
+`source2`. While this is somewhat low-level, when combined with subsequent
+transforms, it gives the author the ability to combine an arbitrary number of
+data sources into any shape they require.
+
+In the following example the author combines the additional text source with the
+compilation:
+
+```csharp
+var combined = context.Sources.AdditionalTexts.Combine(context.Sources.Compilation);
+```
+
+The type of combined is an
+`IncrementalValueSource<(AdditionalText, IEnumerable<Compilation>)>`. For each
+additional text, there is a tuple with the text, and the batched data of the
+compilation source. As the author knows that there is only ever a single
+compilation, they are free to transform the data to select the single
+compilation object:
+
+```csharp
+IncrementalValueSource<(AdditionalText, Compilation)> transformed = combined.Transform(pair => (pair.Item1, pair.Item2.Single()));
+```
```

Comment:
the .Single here is super sucky.  I wonder if we can have a better Combine that reflects this.

Reply:
Yeah, that was the original reasoning for the Single/MultiValueSource stuff, you got the 'single' for free via the type system. We could have a `CombineWithSingle()` that just throws at runtime if the second argument isn't a single thing.